### ISSUE TITLE
Installation Instructions for Jupyter

### ISSUE THREAD

**Issue by ZelphirKaltstahl** [mattpap/IScala#37]
I found your repository by clicking a link in the Jupyter documentation. Afaik IPython Notebook is obsolete and work is done on Jupyter now. Could you add clear working instructions for how to use your kernel in Jupyter?

What I mean is that the workflow to get it running on IPython doesn't work for Jupyter. For example there is no

> jupyter profile create iscala

because jupyter doesn't know the profile command. So I tried using the IPython command and then making jupyter use the configuration that command creates using:

> jupyter notebook --profile iscala

However, that command doesn't exist either. So I looked up the available commands and it turns out there is

> jupyter --config="..."

This is what I used then. I specified the path to your IScala.jar in that config file created by IPython. However, when I want to create a new notebook it only shows the python3 kernel. I thought "Maybe it simply doesn't know that it isn't python, but Scala instead." and proceeded. Then it connects to the kernel when the notebook is shown. When it connects your kernel dies.

TLDR: I couldn't get your kernel to run with Jupyter. Can you please provide detailed instructions on how to achieve that?

**Comment by frgomes** [comment]
Hello,

I've written a gist which may be of your interest: https://gist.github.com/frgomes/24c1923f2c9f14a27b14
It works on Debian and Kubuntu. Potentially works fine in other Debian derivatives as well.

_These instructions above are not related to IScala, though._

I was using IScala before and I simply tried to open IPython/IScala notebooks from Jupyter.
Jupyter migrates the format and opens notebooks but obviously IScala specific stuff is not understood by Jupyter. You basically need to replace instructions involving repositories and dependencies.

Using the solution I pointed out (_which is not IScala, but another thing!_) you can define repositories and dependencies like so:

```
load.resolver("jcenter" at "https://jcenter.bintray.com/")
load.ivy("info.folone" %% "poi-scala" % "0.15")
```

You may also find troubles displaying images, I mean: your notebooks working fine with IScala may not display images properly using the other solution I pointed out. This is a topic I'm currently investigating how this could be solved.

I hope it helps.

**Comment by frgomes** [comment]
By the way, during the migration process, kernel IScala is not understood by Jupyter.
You can simply tell Jupyter to change the kernel. After that, you have to change instructions related to repositories and loading dependencies. For example:

    load.resolver("jcenter" at "https://jcenter.bintray.com/")
    load.ivy("info.folone" %% "poi-scala" % "0.15")

**Comment by ZelphirKaltstahl** [comment]
I should've mentioned in the OP: I created a virtual environment using anaconda3 and python3.5, selecting the jupyter package. This is the virtual environment I wanted to add the Scala kernel to.

I'd like to get it running in that virtual environment, so that it doesn't add anything to the system. Is there an easy way to do that? Maybe changing some paths or something? I couldn't find any packages on PyPI for Jupyter kernels.

**Comment by frgomes** [comment]
As you do, I also do prefer installing things locally, separated from the default/standard Linux distribution.

The script I pointed out does that, I mean: it installs a virtual environment first and then installs IPython _inside_ the virtual environment. Also, I install Java, Scala and other tools under a folder I manage myself. So, the system contains only the Linux distribution itself.

I do not have experience with Anaconda. I prefer to have full control of what I'm doing. Anaconda, despite useful, leads me to depend on a third party. So, I cannot help you in regards to Anaconda.

**Comment by ZelphirKaltstahl** [comment]
The Anaconda venvs can be used just like usual ones. Anaconda itself is free software according to their website. You can still use pip to install stuff from PyPI, activate and deactivate them and list all installed packages. Anaconda uses standard venvs, which it puts into its own directory in a subdirectory called envs.

Anyway, I can try with normal venvs as well.

I guess

> /opt/developer/

is that directory you mentioned, which you manage yourself?

Couldn't I simply replace that self managed directory with the path of my virtual environments lib directory, if I wanted to contain everything jupyter in its virtualenv?

The second download, the

> jupyter-scala_2.11.6-0.2.0-SNAPSHOT.tar.xz

Is it an installer or is it just the Scala kernel?

> ipython notebook

I could simply use

> jupyter notebook

At that time, right?

What does postactivate actually do?

**Comment by frgomes** [comment]
Yes, `/opt/developer` is the directory I install things by hand, and you can replace by whatever you wish. I just uncompress packages in there, such as JDK, Scala, Maven, SBT, etc. Then, my virtual environment makes references to these folders I create by hand. More details here: https://gist.github.com/frgomes/6f5577bc1898445dd1f8

I'm using `ipython notebook` and it works just fine, starting jupyter.
Thanks for pointing out to `jupyter notebook`... I was not aware of that. :-)

**Comment by lJoublanc** [comment]
Here's an alternative. I got this working by following the [jupyter kernel docs](https://jupyter-client.readthedocs.io/en/latest/kernels.html#kernelspecs).

First, clone this repository, and run the sbt commands:
1. `compile`
2. `export runtime:fullClasspath`

This will fetch all the necessary libraries to your local ivy directory, and compile the code. The output from the second command you will use below.

Second, create the file `%APPDATA%\jupyter\kernels\IScala\kernel.json` for windows, (or see the above link for *nix). It should have the following contents:

```
{
 "argv": ["java",
          "-cp", "C:\\Users\\joebloggs\\local\\IScala\\target\\scala-2.11\\classes;C:\\Users\\joebloggs\\local\\IScala\\core\\target\\scala-2.11\\classes; ...",
          "org.refptr.iscala.IScala",
          "--connection-file", "{connection_file}", "--parent"],
 "display_name": "IScala",
 "language": "scala"
}

```
where the class path (string following `-cp` above) is the output of `sbt "export runtime:fullClasspath"`.

This should work, and you'll have the latest version.
Note that this runs from your compiled code, rather than using the `IScala.jar` from from the releases page.
